// @open-dy/open_api_credential/dist/client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = void 0;
/*
Copyright 2024 ByteDance Ltd. and/or its affiliates.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/
const axios_1 = __importDefault(require("axios"));
class Config {
    constructor(map) {
        if (map) {
            for (let key in map) {
                if (this.hasOwnProperty(key)) {
                    this[key] = map[key];
                }
            }
        }
    }
}
exports.Config = Config;
class Client {
    // // code 和 businessToken 做一个映射  
    // private businessTokenMap: Record<string, BusinessToken> = {};
    // // code 和 userAcessToken 做一个映射
    // private accessTokenMap: Record<string, UserAccessToken> = {};
    constructor(config) {
        this.clientKey = config.clientKey;
        this.clientSecret = config.clientSecret;
    }
    // private getBusinessKey(code: string, scope: string): string {
    //   return scope + '-' + code;
    // }
    // private setBusinessToken(code: string, scope: string, businessToken: BusinessToken) {
    //   this.businessTokenMap[this.getBusinessKey(code, scope)] = { ...this.businessTokenMap[this.getBusinessKey(code, scope)], ...businessToken };
    // }
    // private setUserAccessToken(code: string, accessToken: UserAccessToken) {
    //   this.accessTokenMap[code] = { ...this.accessTokenMap[code], ...accessToken };
    // }
    getClientKey() {
        return this.clientKey;
    }
    getClientSecret() {
        return this.clientSecret;
    }
    formatErrorMessage(errorCode, errorMessage) {
        return `biz code not 0, code=${errorCode},message=${errorMessage}`;
    }
    async getClientTokenHttp() {
        var _a, _b, _c, _d, _e;
        let res = await axios_1.default.post('https://open.douyin.com/oauth/client_token/', {
            grant_type: 'client_credential',
            client_key: this.clientKey,
            client_secret: this.clientSecret
        }, {
            headers: {
                'Content-Type': 'application/json',
            }
        });
        const result = res.data;
        if (((_a = result === null || result === void 0 ? void 0 : result.data) === null || _a === void 0 ? void 0 : _a.error_code) === 0) {
            this.clientToken = (_b = result === null || result === void 0 ? void 0 : result.data) === null || _b === void 0 ? void 0 : _b.access_token;
            this.clientTokenExpires = Date.now() + ((_c = result === null || result === void 0 ? void 0 : result.data) === null || _c === void 0 ? void 0 : _c.expires_in) * 1000;
            return { accessToken: this.clientToken, expiresIn: this.clientTokenExpires };
        }
        else {
            throw new Error(this.formatErrorMessage((_d = result === null || result === void 0 ? void 0 : result.data) === null || _d === void 0 ? void 0 : _d.error_code, (_e = result === null || result === void 0 ? void 0 : result.data) === null || _e === void 0 ? void 0 : _e.description));
        }
    }
    setClientTokenExpires(expires) {
        this.clientTokenExpires = Date.now() + expires;
    }
    async getClientToken() {
        try {
            if (!this.clientToken) {
                return this.getClientTokenHttp();
            }
            else if (Date.now() > this.clientTokenExpires) {
                return this.getClientTokenHttp();
            }
            else {
                return { accessToken: this.clientToken, expiresIn: this.clientTokenExpires };
            }
        }
        catch (error) {
            throw error;
        }
    }
    async getAccessTokenHttp() {
        var _a, _b;
        let res = await axios_1.default.post('https://developer.toutiao.com/api/apps/v2/token', {
            grant_type: 'client_credential',
            appid: this.clientKey,
            secret: this.clientSecret
        }, {
            headers: {
                'Content-Type': 'application/json',
            }
        });
        const result = res.data;
        if ((result === null || result === void 0 ? void 0 : result.err_no) === 0) {
            this.accessToken = (_a = result === null || result === void 0 ? void 0 : result.data) === null || _a === void 0 ? void 0 : _a.access_token;
            this.accessTokenExpires = Date.now() + ((_b = result === null || result === void 0 ? void 0 : result.data) === null || _b === void 0 ? void 0 : _b.expires_in) * 1000;
            return { accessToken: this.accessToken, expiresIn: this.accessTokenExpires };
        }
        else {
            throw new Error(this.formatErrorMessage(result === null || result === void 0 ? void 0 : result.err_no, result === null || result === void 0 ? void 0 : result.err_tips));
        }
    }
    setAccessTokenExpires(expires) {
        this.accessTokenExpires = Date.now() + expires;
    }
    async getAccessToken() {
        try {
            if (!this.accessToken) {
                return this.getAccessTokenHttp();
            }
            else if (Date.now() > this.accessTokenExpires) {
                return this.getAccessTokenHttp();
            }
            else {
                return { accessToken: this.accessToken, expiresIn: this.accessTokenExpires };
            }
        }
        catch (error) {
            throw error;
        }
    }
}
exports.default = Client;
