// @open-dy/open_api_util/dist/client.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtendsParameters = exports.FileField = void 0;
/*
Copyright 2024 ByteDance Ltd. and/or its affiliates.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*/
const stream_1 = require("stream");
const $tea = __importStar(require("@alicloud/tea-typescript"));
const os_1 = __importDefault(require("os"));
const crypto_1 = __importDefault(require("crypto"));
const querystring_1 = __importDefault(require("querystring"));
const DEFAULT_USER_AGENT = `ByteDance (${os_1.default.platform()}; ${os_1.default.arch()}) Node.js/${process.version} Core/1.0.1 TeaDSL/1`;
class FileField extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            filename: 'filename',
            contentType: 'contentType',
            content: 'content',
        };
    }
    static types() {
        return {
            filename: 'string',
            contentType: 'string',
            content: 'Readable',
        };
    }
}
exports.FileField = FileField;
class FileFormStream extends stream_1.Readable {
    constructor(form, boundary) {
        super();
        this.form = form;
        this.keys = Object.keys(form);
        this.index = 0;
        this.boundary = boundary;
        this.streaming = false;
    }
    _read() {
        if (this.streaming) {
            return;
        }
        const separator = this.boundary;
        if (this.index < this.keys.length) {
            const name = this.keys[this.index];
            const fieldValue = this.form[name];
            if (typeof fieldValue.filename === 'string' &&
                typeof fieldValue.contentType === 'string' &&
                fieldValue.content instanceof stream_1.Readable) {
                let body = `--${separator}\r\n` +
                    `Content-Disposition: form-data; name="${name}"; filename="${fieldValue.filename}"\r\n` +
                    `Content-Type: ${fieldValue.contentType}\r\n\r\n`;
                this.push(Buffer.from(body));
                this.streaming = true;
                fieldValue.content.on('data', (chunk) => {
                    this.push(chunk);
                });
                fieldValue.content.on('end', () => {
                    this.index++;
                    this.streaming = false;
                    this.push('\r\n');
                });
            }
            else {
                this.push(Buffer.from(`--${separator}\r\n` +
                    `Content-Disposition: form-data; name="${name}"\r\n\r\n` +
                    `${fieldValue}\r\n`));
                this.index++;
            }
        }
        else {
            this.push(Buffer.from(`--${separator}--\r\n`));
            this.push(null);
        }
    }
}
class ExtendsParameters extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            headers: 'headers',
        };
    }
    static types() {
        return {
            headers: { 'type': 'map', 'keyType': 'string', 'valueType': 'string' },
        };
    }
}
exports.ExtendsParameters = ExtendsParameters;
const makeNonce = (function () {
    let counter = 0;
    let last;
    const machine = os_1.default.hostname;
    const pid = process.pid;
    return function () {
        let val = Math.floor(Math.random() * 1000000000000);
        if (val === last) {
            counter++;
        }
        else {
            counter = 0;
        }
        last = val;
        var uid = `${machine}${pid}${val}${counter}`;
        let shasum = crypto_1.default.createHash('md5');
        shasum.update(uid);
        return shasum.digest('hex');
    };
})();
class Client {
    static getBoundary() {
        return 'boundary' + Math.random().toString(16).slice(-12);
    }
    static toFileForm(form, boundary) {
        return new FileFormStream(form, boundary);
    }
    static toBytes(val) {
        return Buffer.from(val);
    }
    static toString(val) {
        return val.toString();
    }
    static parseJSON(val) {
        return JSON.parse(val);
    }
    static async readAsBytes(stream) {
        return await new Promise((resolve, reject) => {
            let onData, onError, onEnd;
            let cleanUp = function () {
                stream.removeListener('error', onError);
                stream.removeListener('data', onData);
                stream.removeListener('end', onEnd);
            };
            let bufs = [];
            let size = 0;
            onData = function (buf) {
                bufs.push(buf);
                size += buf.length;
            };
            onError = function (err) {
                cleanUp();
                reject(err);
            };
            onEnd = function () {
                cleanUp();
                resolve(Buffer.concat(bufs, size));
            };
            stream.on('error', onError);
            stream.on('data', onData);
            stream.on('end', onEnd);
        });
    }
    static async readAsString(stream) {
        let buff = await Client.readAsBytes(stream);
        return Client.toString(buff);
    }
    static async readAsJSON(stream) {
        return Client.parseJSON(await Client.readAsString(stream));
    }
    static getNonce() {
        return makeNonce();
    }
    static getDateUTCString() {
        const now = new Date();
        return now.toUTCString();
    }
    static defaultString(real, defaultValue) {
        return real || defaultValue;
    }
    static defaultNumber(real, defaultValue) {
        return real || defaultValue;
    }
    static toFormString(val) {
        return querystring_1.default.stringify(val);
    }
    static toJSONString(val) {
        if (typeof val === 'string') {
            return val;
        }
        return JSON.stringify(val);
    }
    static empty(val) {
        return !val;
    }
    static equalString(val1, val2) {
        return val1 === val2;
    }
    static equalNumber(val1, val2) {
        return val1 === val2;
    }
    static isUnset(value) {
        if (typeof value === 'undefined') {
            return true;
        }
        if (value === null) {
            return true;
        }
        return false;
    }
    static stringifyMapValue(m) {
        if (!m) {
            return m;
        }
        const result = {};
        for (const [key, value] of Object.entries(m)) {
            if (typeof value === 'undefined' || value === null) {
                continue;
            }
            result[key] = String(value);
        }
        return result;
    }
    static anyifyMapValue(m) {
        return m;
    }
    static assertAsBoolean(value) {
        if (typeof value === 'boolean') {
            return value;
        }
        throw new Error("value is not boolean");
    }
    static assertAsString(value) {
        if (typeof value === 'string') {
            return value;
        }
        throw new Error('value is not string');
    }
    static assertAsBytes(value) {
        if (Buffer.isBuffer(value)) {
            return value;
        }
        throw new Error('value is not bytes');
    }
    static assertAsNumber(value) {
        if (typeof value === 'number') {
            return value;
        }
        throw new Error('value is not number');
    }
    static assertAsInteger(value) {
        if (Number.isInteger(value)) {
            return value;
        }
        throw new Error('value is not integer');
    }
    static assertAsMap(value) {
        if (value && typeof value === 'object' && !Array.isArray(value)) {
            return value;
        }
        throw new Error('value is not object');
    }
    static assertAsArray(value) {
        if (Array.isArray(value)) {
            return value;
        }
        throw new Error('value is not array');
    }
    static getUserAgent(userAgent) {
        if (!userAgent || !userAgent.length) {
            return DEFAULT_USER_AGENT;
        }
        return DEFAULT_USER_AGENT + " " + userAgent;
    }
    static is2xx(code) {
        return code >= 200 && code <= 300;
    }
    static is3xx(code) {
        return code >= 300 && code < 400;
    }
    static is4xx(code) {
        return code >= 400 && code < 500;
    }
    static is5xx(code) {
        return code >= 500 && code < 600;
    }
    static validateModel(m) {
        throw new Error('Un-implemented!');
    }
    static toMap(model) {
        return $tea.toMap(model);
    }
    static async sleep(millisecond) {
        return new Promise(resolve => {
            setTimeout(() => {
                resolve();
            }, millisecond);
        });
    }
    static toArray(input) {
        if (!(input instanceof Array)) {
            return null;
        }
        let ret = [];
        input.forEach((model) => {
            if (!model) {
                return;
            }
            ret.push($tea.toMap(model));
        });
        return ret;
    }
    static assertAsReadable(value) {
        if (value instanceof stream_1.Readable) {
            return value;
        }
        throw new Error('value is not readable');
    }
    static getErrNo(errMsg) {
        var _a, _b;
        if (!!(errMsg === null || errMsg === void 0 ? void 0 : errMsg.err_no)) {
            return errMsg.err_no;
        }
        if (!!(errMsg === null || errMsg === void 0 ? void 0 : errMsg.errcode)) {
            return errMsg.errcode;
        }
        if (!!(errMsg === null || errMsg === void 0 ? void 0 : errMsg.error_code)) {
            return errMsg.error_code;
        }
        if (!!(errMsg === null || errMsg === void 0 ? void 0 : errMsg.errno)) {
            return errMsg.errno;
        }
        if (!!(errMsg === null || errMsg === void 0 ? void 0 : errMsg.error)) {
            return errMsg.error;
        }
        if (!!((_a = errMsg === null || errMsg === void 0 ? void 0 : errMsg.err) === null || _a === void 0 ? void 0 : _a.err_code)) {
            return errMsg.err.err_code;
        }
        if (!!((_b = errMsg === null || errMsg === void 0 ? void 0 : errMsg.extra) === null || _b === void 0 ? void 0 : _b.error_code)) {
            return errMsg.extra.error_code;
        }
        return 0;
    }
    static getErrMsg(errMsg) {
        var _a, _b;
        if (!!(errMsg === null || errMsg === void 0 ? void 0 : errMsg.err_msg)) {
            return errMsg.err_msg;
        }
        if (!!(errMsg === null || errMsg === void 0 ? void 0 : errMsg.err_tips)) {
            return errMsg.err_tips;
        }
        if (!!(errMsg === null || errMsg === void 0 ? void 0 : errMsg.errmsg)) {
            return errMsg.errmsg;
        }
        if (!!((_a = errMsg === null || errMsg === void 0 ? void 0 : errMsg.err) === null || _a === void 0 ? void 0 : _a.err_msg)) {
            return errMsg.err.err_msg;
        }
        if (!!((_b = errMsg === null || errMsg === void 0 ? void 0 : errMsg.extra) === null || _b === void 0 ? void 0 : _b.description)) {
            return errMsg.extra.description;
        }
        return "";
    }
    static getLogid(errMsg) {
        var _a;
        if (!!(errMsg === null || errMsg === void 0 ? void 0 : errMsg.log_id)) {
            return errMsg.log_id;
        }
        if (!!((_a = errMsg === null || errMsg === void 0 ? void 0 : errMsg.extra) === null || _a === void 0 ? void 0 : _a.logid)) {
            return errMsg.extra.logid;
        }
        return "";
    }
    static getErrMessage(msg) {
        let errMsg = JSON.parse(msg);
        return {
            err_no: this.getErrNo(errMsg),
            err_msg: `msg is: ${this.getErrMsg(errMsg)},logid is ${this.getLogid(errMsg)}`,
        };
    }
}
exports.default = Client;
